{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "42d05392",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"../images/logo.png\" alt=\"GML QIMR Logo\">\n",
    "</div>\n",
    "\n",
    "**This R script/tutorial has been generated by Prakrithi from the *Genomics and Machine Learning Lab*:**\n",
    "* **Data used below is currently not published or publically available. For more information about the scripts or for the use of this datasets beyond this tutorial please contact uqacause@uq.edu.au**\n",
    "* **Code and *conda* environments used are available from https://github.com/GMLTestLab/qimr-teaching**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e610e71d-7bb3-4a6e-a653-b9db8992fde5",
   "metadata": {},
   "source": [
    "# TABLE OF CONTENTS\n",
    "\n",
    "1. [Introduction to CNV profiling](#Introduction)\n",
    "2. [scRNA-seq Data Loading](#Data-Loading)\n",
    "3. [CopyKAT - Identifying Malignant melanocytes and identifying patient with the tumor](#CopyKAT)\n",
    "4. [InferCNV](#InferCNV)\n",
    "5. [Consistency of prediction between inferCNV and CopyKAT](#Evaluation)\n",
    "6. [Analyzing subclones in spatial data using CopyKAT](#spatial)\n",
    "7. [Summary](#SUMMARY)\n",
    "\n",
    "\n",
    "This is just a tutorial and not a hands-on notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39a527f6-1852-4ba6-8823-a12fe5024cb0",
   "metadata": {},
   "source": [
    "# CNV Profiling of Single cell and Spatial RNA sequencing data to identify Tumor cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "602fad30-ecf6-44b7-9bfc-0991a0d60f69",
   "metadata": {},
   "source": [
    "### What are CNVs?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce4460f1-dce6-4cca-bf82-b4616e6a529c",
   "metadata": {},
   "source": [
    "A Copy number variation (abbreviated as CNV) refers to an instance in which the number of copies of a specific DNA segment   varies among different individuals' genomes. These variations can involve deletions or duplications of segments of the genome and can range from a few kilobases to several megabases in size. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa2bad04-33ed-4e28-b593-c101d4ce19d2",
   "metadata": {},
   "source": [
    "### How are CNVs related to cancer?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be79756c-f7f8-438d-be5e-e3e8af7c63de",
   "metadata": {},
   "source": [
    "- Oncogene Amplification\n",
    "- Tumor Suppressor Gene Deletion\n",
    "- Genomic Instability"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59be816accfcdeb7",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<img src=\"images/CNVs.jpeg\" alt=\"CNVs\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6599565-46fe-4adf-9ca1-2c1985e7235f",
   "metadata": {},
   "source": [
    "### How can we make use of this DNA profile information for RNA-seq data?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2109e8d8-f481-4363-95e5-278050091912",
   "metadata": {},
   "source": [
    "The distinction of malignant from non-malignant cells is a critical step in the follow-up analysis of scRNA-seq tumor datasets. The basic idea to solve such a problem relies on estimating common copy number alterations that characterize transformed cells. The copy number profiles are obtained by considering the gene expression profiles of each cell as a function of the genomic coordinates. The moving average smoothing of the gene expression function is then clustered in malignant and non-malignant cells. The underlying logic for calculating DNA copy number events from RNAseq data is that gene expression levels of many adjacent genes can provide depth information to infer genomic copy number in that region."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6091859-6b56-43a8-b4bc-7fc24807f816",
   "metadata": {},
   "source": [
    "## scRNA-Seq Melanoma dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ff9f7a9-8583-4dc4-b4be-2c9f3b1c76c9",
   "metadata": {},
   "source": [
    "An in-house Melanoma dataset consisting of three samples - One with a malignant tumor and the rest with Dysplastic Nevi (unusual mole that has the potential to turn malignant) has been used to demonstrate the application of CNVs in identifying cancerous cells."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d106a24c-2dfe-4ac3-8dd7-bf60fb7a28cf",
   "metadata": {},
   "source": [
    "### Load necessary libraries and set data directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2589401-a491-4273-82c2-910b0cf00883",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "data_dir <- R.utils::getAbsolutePath('../data/infercnv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c812b090-f5bc-470d-a0cf-231fe4471320",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "## Load the integrated Melanoma dataset .rds file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7de4dceb-3af2-427e-97ba-fdb4c48cb9a9",
   "metadata": {},
   "source": [
    "Load the seurat object (.rds file converted from anndata object) as Mel_scRNA. <br> Mel_scRNA@meta.data$cell_type has the annotated cell type information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43bece69-4121-491f-a661-39b6a906a67a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Mel_scRNA<-readRDS(glue::glue(\"{data_dir}/Mel_3samples_75pcs.rds\"))\n",
    "Mel_scRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37c9161-02c3-4f1d-9c2d-5e5e385ad257",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "Idents(Mel_scRNA)<-Mel_scRNA@meta.data$cell_type\n",
    "DimPlot(Mel_scRNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5876ef2d963a2751",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<img src=\"images/umap.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "067dd204-d82f-43fc-9fe9-d1e69299a19c",
   "metadata": {},
   "source": [
    "**Melanocytes** are the cells from which \n",
    "Melanoma arises. Not all the melanocytes are cancerous. It is necessary to correctly identify the malignant cells, to resolve clonal substructure within the tumor and to identify the patient sample they come from for accurate diagnosis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9422e809-5a84-471a-b634-81d7dec8ea9e",
   "metadata": {},
   "source": [
    "<a id=\"CopyKAT\"></a>\n",
    "## CopyKAT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39a6a8ab-ed41-4bfc-8b45-e7203b222d16",
   "metadata": {},
   "source": [
    "**CopyKAT (Copy Number Karyotyping of Aneuploid Tumors)** is a tool to identify CNVs from high dimensional RNA-Seq dataset like scRNA-Seq and can be applied to spatial transcriptomics data as well.\n",
    "\n",
    "It is a reference-free method i.e. It does not require prior cell type annotations defining normal cell profiles."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da220449-adbf-4e08-b376-573f62a639d1",
   "metadata": {},
   "source": [
    "**Steps involved**\n",
    "\n",
    "- **step1**: read and filter data. **Input:** Gene - Barcode count matrix\n",
    "- **step 2**: annotations of genes and ordering them by gene coordinates\n",
    "- **step 3**: Freeman–Tukey transformation is performed to stabilize variance, followed by polynomial dynamic linear modeling (DLM) to smooth the outliers in the single-cell UMI counts\n",
    "- **step 4**:  detect a subset of diploid cells with high confidence to infer the copy number baseline values of the normal 2N cells. To do this, single cells are pooled into several small hierarchical clusters and estimate the variance of each cluster using a Gaussian mixture model (GMM). The cluster with minimal estimated variance is defined as the ‘confident diploid cells’ by following a strict classification criterion.\n",
    "- **step 5**: segmentation - Relative gene expression values in single cells are used for MCMC segmentation, and segments are merged by KS testing.\n",
    "- **step 6**: Prediction: Aneuploid tumor and normal cell clusters are classified using normal cell enrichment and GMM distribution tests. The clonal substructure of tumor cells is delineated by clustering, and subclones are used for downstream analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47c2e8e3-a108-4bfe-a510-dce52ce0b7c5",
   "metadata": {},
   "source": [
    "![copyKAT](images/41587_2020_795_Fig1_HTML.png)\n",
    "\n",
    "Nature Biotechnology (Nat Biotechnol) ISSN 1546-1696 (online) ISSN 1087-0156 (print)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03cb77bf-2993-453f-89a6-476116749cb0",
   "metadata": {},
   "source": [
    "More details at https://github.com/navinlabcode/copykat\n",
    "\n",
    "Publication: Gao R et al. (2021). Nat Biotechnol. doi:10.1038/s41587-020-00795-2."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "805e5f07-c19e-43c5-ba60-d433ebb67402",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "### copyKAT installation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b345c9d6-5564-4013-8163-8d98f5757a89",
   "metadata": {},
   "source": [
    "Uncomment the lines below to install copyKAT and load the library once installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "040a8b57-fe41-428c-8e4e-4736307cc906",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(copykat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aba5b945-cd88-4d77-b970-41997028ba4d",
   "metadata": {},
   "source": [
    "## Run CopyKAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31cfc609-a724-4023-aae4-0d215a895643",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "exp.rawdata<-read.csv(glue::glue(\"{data_dir}/MPSs_count_mat.txt\",sep=\"\\t\"))\n",
    "rownames(exp.rawdata)<-exp.rawdata$GENE\n",
    "exp.rawdata<-exp.rawdata[,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9788d364-a531-4b73-9c3d-b4b6d54a0816",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "copykat.test <- copykat(rawmat=exp.rawdata, id.type=\"S\", ngene.chr=5, win.size=25, KS.cut=0.1, sam.name=\"test\",distance=\"euclidean\", \n",
    "                        norm.cell.names=\"\",output.seg=\"FLASE\", plot.genes=\"TRUE\", genome=\"hg20\",n.cores=1) #hg20 built-in copkat is the hg38 genome build coordinatess\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad812dce-6f85-47d3-bf77-465e3546999d",
   "metadata": {},
   "source": [
    "### Read CopyKAT predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6c5032-b9f1-49be-8cd6-7d410e93693b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "p<-read.csv(glue::glue(\"{data_dir}/test_copykat_prediction.txt\"), sep=\"\\t\",header=TRUE)\n",
    "head(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9601b6b8-90c1-4b9b-8ed7-3dcd46e08bc3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "colnames(Mel_scRNA)<-gsub(\"-1\",\".1\",colnames(Mel_scRNA))\n",
    "seurat_colnames<-colnames(Mel_scRNA)\n",
    "rownames(p)<-p$cell.names\t\n",
    "p<-p[seurat_colnames,]\n",
    "# Assuming infercnv_sub_groupings is your DataFrame and seurat_colnames contains the column names from your Seurat object\n",
    "\n",
    "# Find the order of the BC values in the seurat_colnames\n",
    "bc_order <- match(p$cell.names, seurat_colnames)\n",
    "\n",
    "# Reorder the columns of infercnv_sub_groupings based on the order\n",
    "p_reordered <- p[order(bc_order), ]\n",
    "\n",
    "# View the reordered DataFrame\n",
    "dim(p_reordered)\n",
    "head(p_reordered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1bc00df-9c92-4c42-ad5d-9dd8979f9526",
   "metadata": {},
   "source": [
    "## Identify Malignant Tumor cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "118962dc-07b4-487c-9483-d0b77888aa47",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Mel_scRNA@meta.data$copykat_pred<-p$copykat.pred\n",
    "#Mel_scRNA@meta.data$copykat_aneuploid <- ifelse(Mel_scRNA@meta.data$copykat_pred == \"aneuploid\", \"Aneuploid\", \"\")\n",
    "aneuploid_cells <- rownames(Mel_scRNA@meta.data)[Mel_scRNA@meta.data$copykat_pred == \"aneuploid\"]\n",
    "\n",
    "DimPlot(Mel_scRNA, group.by = \"copykat_pred\")\n",
    "DimPlot(Mel_scRNA, group.by = \"copykat_pred\", cells.highlight = aneuploid_cells, cols.highlight = \"red\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "064f4401-367d-4e9a-99da-5ffa4c38a6d5",
   "metadata": {},
   "source": [
    "## Identify the sample with malignant tumor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "114713c1-ba0b-4795-aa3d-4064c25eb5ec",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 6)\n",
    "DimPlot(Mel_scRNA, group.by = \"copykat_pred\", split.by = \"orig.ident\")\n",
    "DimPlot(Mel_scRNA, group.by = \"copykat_pred\", cells.highlight = aneuploid_cells,cols.highlight = \"red\", split.by = \"orig.ident\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbfea3eb-bd98-491c-84a0-d0a90043afd6",
   "metadata": {},
   "source": [
    "## InferCNV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9ceab9c-b01d-49c6-8e88-6ac8beadcfc3",
   "metadata": {},
   "source": [
    "Another tool many used to analyze CNV profiles of tumor cells and to classify subclones\n",
    "\n",
    "- InferCNV uses a corrected moving average of gene expression data to determine CNV profiles. \n",
    "- Genes are sorted by absolute genomic position. That is, they are first ordered by chromosome, and then by genomic start position within the chromosome. averaging out the expression of genomically adjacent genes removes gene-specific expression variability and yields profiles that reflect chromosomal copy number variations.\n",
    "- To further refine the CNV profile of tumor cells, InferCNV constructs the CNV profile of a known normal sample, and then for each gene and each cell, the normal sample is subtracted from the tumor sample to determine the final tumor CNV profile.\n",
    "\n",
    "More Information at https://github.com/broadinstitute/infercnv/wiki\n",
    "\n",
    "![inferCNV](https://github.com/broadinstitute/infercnv/wiki/images/InferCNV_procedure.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a22b8af-5dc6-4627-9b82-96bc559261ab",
   "metadata": {},
   "source": [
    "### Data requirements"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08b1605f-1c29-4b17-958e-d3a6d093ddf3",
   "metadata": {},
   "source": [
    "inferCNV requires:\n",
    "\n",
    "- a raw counts matrix of single-cell RNA-Seq expression\n",
    "- an annotations file which indicates which cells are tumor vs. normal.\n",
    "- a gene/chromosome positions file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d157280f-0288-444d-ba0e-fe9a7de26024",
   "metadata": {},
   "source": [
    "Detailed documentation: https://www.bioconductor.org/packages/release/bioc/manuals/infercnv/man/infercnv.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "777fdc3f-be51-4324-9f5d-853fe84c9560",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95d5bfdc-7e82-4ad1-b2cc-ceb9e80d8915",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(infercnv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "714bbf69-9a3c-46d1-9608-0887f596624d",
   "metadata": {},
   "source": [
    "### Preparing metadata file containing cell-type information from previously loaded seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35c6f749-5e4b-4503-8cc8-93575e27d9bd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(\"Mel_scRNA.assay.version\" = \"v3\")\n",
    "infercnv_meta <- data.frame(V1 = Mel_scRNA@meta.data$cell_type)\n",
    "rownames(infercnv_meta)<-rownames(Mel_scRNA@meta.data)\n",
    "head(infercnv_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c655868-9dcb-4df2-bfca-67c05c317562",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "unique(Mel_scRNA@meta.data$Level1_res1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ad1a93-8b89-41d3-9a7f-75f8107729aa",
   "metadata": {},
   "source": [
    "### Create InferCNV object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be974ecf-6efc-427b-be97-d6bc1b6858a3",
   "metadata": {},
   "source": [
    "You can set the reference 'Normal' cell-types as a list using ref_group_names. Or you could set it to NULL. <br>\n",
    "Either way inferCNV identifies aneuploid cells and infers subclones using predicted CNV profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "543b43c2-2036-4007-a2f3-a329bba51016",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "infercnv_obj = CreateInfercnvObject(\n",
    "  raw_counts_matrix=GetAssayData(Mel_scRNA, slot=\"counts\"),\n",
    "  annotations_file=infercnv_meta,\n",
    "  delim=\"\\t\",\n",
    "  gene_order_file=glue::glue(\"{data_dir}/hg38_gencode_v27.txt\"),\n",
    "  ref_group_names=NULL) #c(\"KC\",\"Immune\",\"Fibroblast\",\"Endothelial cell\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c52ba298-929f-46ba-ae7c-8af98df553e6",
   "metadata": {},
   "source": [
    "### Run InferCNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91759be4-9e48-4735-a5cf-8bb5814dac82",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "out_dir = tempfile()\n",
    "infercnv_obj_default = infercnv::run(\n",
    "    infercnv_obj,\n",
    "    cutoff=1, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics\n",
    "    out_dir=glue::glue(\"{data_dir}/all\"),\n",
    "    cluster_by_groups=TRUE, \n",
    "    plot_steps=FALSE,\n",
    "    denoise=TRUE,\n",
    "    HMM=FALSE,\n",
    "    no_prelim_plot=TRUE,\n",
    "    png_res=60\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "902c827c-654d-49a0-b6b0-71ff039b89a1",
   "metadata": {},
   "source": [
    "![infercnv_hm_ref](images/infercnv.png)\n",
    "\n",
    "When you specify reference cell types, they are clustered separately. If not they are still clustered separately but are given cluster numbers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd24db9-5cff-4ab5-bfcc-072d07f30235",
   "metadata": {},
   "source": [
    "In both cases, the aneuploid cells are subclustered. The diploid cells are excluded in the subcluster-prediction file infercnv_subclusters.observation_groupings.txt. Pick the cells following a prefix 1_ in the HeatMap since this is the largest aneuploid cluster (Melanocytes). The other inclusions could be artifacts or false predictions and might need further investigation. They could also be validated by comparing predictions from CopyKAT and choose the cells with consistent aneuploid predictions across both tools.\n",
    "\n",
    "![infercnv_subclusters](images/QIMR_infercnv_subclusters.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e056bdfa-ad2b-492f-a1e5-365c15881dfb",
   "metadata": {},
   "source": [
    "Alternately, you could calculate CNV scores <br> **scores=apply(infercnv_obj@expr.data,2,function(x){ sum(x < 0.95 | x > 1.05)/length(x) })**\n",
    " <br> and plot the distribution of a gene. If you see a binormal distribution, you could choose the cut-off based on where the first peak drops to choose aneuploid cells."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e274f8-6639-495f-b4d6-bb1547ca6801",
   "metadata": {},
   "source": [
    "Example:\n",
    "![hist](images/infercnv_hist.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f9fe077-91ee-4890-8ba6-e7ae6ac3627b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "infercnv_sub_groupings<-read.csv(glue::glue(\"{data_dir}/all/infercnv_subclusters.observation_groupings.txt\"), sep=\" \",header=TRUE)\n",
    "tail(infercnv_sub_groupings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b27b2a3-2ad7-4f9c-ad4a-33d98aeececb",
   "metadata": {},
   "source": [
    "infercnv_subclusters.observation_groupings.txt contains the sub-clusters of the Aneuploid cells. 284 cells are predicted aneuploid and are further classified into 12 clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f58b348-ba28-4598-95de-964e329d94d8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "unique(infercnv_sub_groupings$Annotation.Group)\n",
    "dim(infercnv_sub_groupings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b36bee89-6d46-4006-ac68-44c88dcd1ae2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rownames(infercnv_sub_groupings)<-infercnv_sub_groupings$BC\n",
    "\n",
    "# Find the order of the BC values in the seurat_colnames\n",
    "bc_order <- match(rownames(infercnv_sub_groupings), colnames(Mel_scRNA))\n",
    "\n",
    "# Reorder the columns of infercnv_sub_groupings based on the order\n",
    "infercnv_sub_groupings_reordered <- infercnv_sub_groupings[order(bc_order), ]\n",
    "\n",
    "# View the reordered DataFrame\n",
    "head(infercnv_sub_groupings_reordered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e169ab8-8444-4c08-a564-c2374b4676f7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# Create a new column in Mel_scRNA@meta.data with empty values\n",
    "Mel_scRNA@meta.data$infercnv_pred <- \"\"\n",
    "\n",
    "# Get the row names of both data frames\n",
    "mel_row_names <- rownames(Mel_scRNA@meta.data)\n",
    "infercnv_row_names <- rownames(infercnv_sub_groupings)\n",
    "\n",
    "# Find the matching row names\n",
    "matching_row_names <- intersect(mel_row_names, infercnv_row_names)\n",
    "\n",
    "# Update the new column with \"Aneuploid\" for matching rows\n",
    "Mel_scRNA@meta.data[matching_row_names, \"infercnv_pred\"] <- \"Aneuploid\"\n",
    "\n",
    "# Check the updated meta.data\n",
    "head(Mel_scRNA@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78c2c3c9-5064-4d0e-adf1-ff47ac043e96",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "infercnv_aneuploid_cells <- rownames(Mel_scRNA@meta.data)[Mel_scRNA@meta.data$infercnv_pred == \"Aneuploid\"]\n",
    "DimPlot(Mel_scRNA, group.by = \"infercnv_pred\", cells.highlight = infercnv_aneuploid_cells, cols.highlight = \"red\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76464da0-9057-4fc1-b987-b62c1af3b537",
   "metadata": {},
   "source": [
    "<a id=\"Evaluation\"></a>\n",
    "\n",
    "## Consistency of prediction between inferCNV and CopyKAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8210c66-3424-4650-afac-2c5302fe2f65",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create a new column in Mel_scRNA@meta.data with empty values\n",
    "Mel_scRNA@meta.data$copykat_infercnv_consistent <- \"\"\n",
    "\n",
    "# Update the new column with \"Aneuploid\" where both conditions are met\n",
    "Mel_scRNA@meta.data$copykat_infercnv_consistent[Mel_scRNA@meta.data$infercnv_pred == \"Aneuploid\" & Mel_scRNA@meta.data$copykat_pred == \"aneuploid\"] <- \"Aneuploid\"\n",
    "\n",
    "# Highlight the cells that are consistent between infercnv and copykat predictions\n",
    "consistent_cells <- rownames(Mel_scRNA@meta.data)[Mel_scRNA@meta.data$copykat_infercnv_consistent == \"Aneuploid\"]\n",
    "DimPlot(Mel_scRNA, group.by = \"copykat_infercnv_consistent\", cells.highlight = consistent_cells, cols.highlight = \"red\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4510551-b58d-4e13-894e-fbd42ebbbbae",
   "metadata": {},
   "source": [
    "The red cells indicate the cells predicted aneuploid with both copyKAT and inferCNV which indicates a 77.89% consistency. <br> 15% of the Melanoytes are predicted malignant with inferCNV, 10%  with copyKAT,  and 8% with both tools."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a4af19f-85a7-47cb-a940-44d4990ec035",
   "metadata": {},
   "source": [
    "<a id=\"spatial\"></a>\n",
    "\n",
    "## Analyzing subclones in SPATIAL data using CopyKAT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f7ba124-f093-41b1-ab96-ed4b3192472d",
   "metadata": {},
   "source": [
    "These tools can be applied on spatial data to visualize subclones. The following demonstrates the use of copyKAT on a public 10X Visium Melanoma dataset as the cell type annotations are unknown.\n",
    "\n",
    "Dataset link https://www.10xgenomics.com/datasets/human-melanoma-if-stained-ffpe-2-standard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d3acff1-8ffb-4a33-bb22-10edd0340e08",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(copykat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "906464a4-77c0-4693-9bbb-0217965a891d",
   "metadata": {},
   "source": [
    "### Extract and use counts from the .h5 file for copyKAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "870e326f-dd11-4955-a4ae-e39ef275dd71",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "exp.rawdata <- read.csv(glue::glue(\"{data_dir}/gene_counts_mat.txt\"),sep=\"\\t\", header=TRUE) #gene-expression matrix\n",
    "rownames(exp.rawdata)<-exp.rawdata$GENE\n",
    "exp.rawdata<-exp.rawdata[,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25afacaa-5ef5-4a2d-a410-f46b0db48f7f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    },
    "is_executing": true
   },
   "outputs": [],
   "source": [
    "copykat.test <- copykat(rawmat=exp.rawdata, id.type=\"S\", ngene.chr=5, win.size=25, KS.cut=0.1, sam.name=\"test\",\n",
    "                        distance=\"euclidean\", norm.cell.names=\"\",output.seg=\"FLASE\", plot.genes=\"TRUE\", \n",
    "                        genome=\"hg20\",n.cores=1) #hg20 built-in copkat is the hg38 coords"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb165ee6-4a2c-436d-b583-e05b5d2ea290",
   "metadata": {},
   "source": [
    "**Output Heatmap showing clustering of Diploid and aneuploid cells base on inferred CNV profiles**\n",
    "\n",
    "![HM](images/QIMR_spatial_test_copykat_heatmap.jpeg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf7e7d0-ff2d-4310-9186-3ea72be41d49",
   "metadata": {},
   "source": [
    "### Extract subclone annotations based on visualizing the output Heatmap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "345ce564-8820-459b-a310-b466537c8cb4",
   "metadata": {},
   "source": [
    "We clearly see the tree splirring into three branches, broadly indicating 4 subclones. So, we cut the tree at '3' and label the subclones to use for further visualization on the tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e3d2e14-db3d-4f0f-a70c-b4e48c052de6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## extract ploidy \n",
    "pred.test <- data.frame(copykat.test$prediction)\n",
    "pred.test <- pred.test[-which(pred.test$copykat.pred==\"not.defined\"),]  ##remove undefined cells\n",
    "CNA.test <- data.frame(copykat.test$CNAmat)\n",
    "\n",
    "### define sub-clones\n",
    "tumor.cells <- pred.test$cell.names[which(pred.test$copykat.pred==\"aneuploid\")]\n",
    "tumor.mat <- CNA.test[, which(colnames(CNA.test) %in% tumor.cells)]\n",
    "hcc <- hclust(parallelDist::parDist(t(tumor.mat),threads =4, method = \"euclidean\"), method = \"ward.D2\")\n",
    "hc.umap <- cutree(hcc,3)\n",
    "# write.table(hc.umap,\"3subclones_BCs.csv\", sep=\",\", quote=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a98a487-c212-4c26-b95c-740ecb6eff10",
   "metadata": {},
   "source": [
    "#### The .h5 file has all information needed to visualize the tissue section and associated data\n",
    "\n",
    "We need to add the subclone information into a new column of the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7030a15d-fa4a-4b8a-a502-628e98403003",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "spatial_data<-Load10X_Spatial(glue::glue(\"{data_dir}/Mel_spatial\"),\n",
    "                    filename = \"CytAssist_FFPE_Human_Skin_Melanoma_filtered_feature_bc_matrix.h5\",\n",
    "                    assay = \"RNA\",\n",
    "                    slice = \"slice1\",\n",
    "                    filter.matrix = TRUE,\n",
    "                    to.upper = FALSE,\n",
    "                    image = NULL\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a93e2c-b2c5-4c59-bc5c-31e82eced13e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# normalize data\n",
    "spatial_data[[\"percent.mt\"]] <- PercentageFeatureSet(spatial_data, pattern = \"^MT-\")\n",
    "spatial_data <- NormalizeData(spatial_data, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
    "spatial_data <- FindVariableFeatures(spatial_data, selection.method = \"vst\", nfeatures = 2000)\n",
    "\n",
    "# scale and run PCA\n",
    "spatial_data <- ScaleData(spatial_data, features = rownames(spatial_data))\n",
    "spatial_data <- RunPCA(spatial_data, features = VariableFeatures(object = spatial_data))\n",
    "\n",
    "# Check number of PC components (we selected 10 PCs for downstream analysis, based on Elbow plot)\n",
    "#ElbowPlot(spatial_data)\n",
    "\n",
    "# cluster and visualize\n",
    "spatial_data <- FindNeighbors(spatial_data, dims = 1:30)\n",
    "spatial_data <- FindClusters(spatial_data, resolution = 0.8)\n",
    "spatial_data <- RunUMAP(spatial_data, dims = 1:30)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c016034-f4dc-40e8-b5a4-aeefbfe29d52",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "### Visualizing the copyKAT predictions to identify tumor region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "805386db-a991-44e8-af42-4b3155be1949",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "p<-read.csv(glue::glue(\"{data_dir}/Mel_spatial/test_copykat_prediction.txt\"), sep=\"\\t\",header=TRUE)\n",
    "colnames(spatial_data)<-gsub(\"-1\",\".1\",colnames(spatial_data))\n",
    "spatial_data_sub<-subset(spatial_data, cells = p$cell.names)\n",
    "\n",
    "seurat_colnames<-colnames(spatial_data_sub)\n",
    "rownames(p)<-p$cell.names\t\n",
    "p<-p[seurat_colnames,]\n",
    "\n",
    "# Find the order of the BC values in the seurat_colnames\n",
    "bc_order <- match(p$cell.names, seurat_colnames)\n",
    "\n",
    "# Reorder the columns of infercnv_sub_groupings based on the order\n",
    "p_reordered <- p[order(bc_order), ]\n",
    "\n",
    "spatial_data_sub@meta.data$copykat_pred<-p$copykat.pred\n",
    "SpatialDimPlot(spatial_data_sub, group.by = \"copykat_pred\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e46604-dc6c-4d65-83a0-ba94367f447d",
   "metadata": {},
   "source": [
    "### Adding subclone information to the spatial object and visualizing subclones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5996a500-fbcd-478b-80d9-f1a10dbce7a1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "c<-read.csv(glue::glue(\"{data_dir}/3subclones_BCs.csv\"))\n",
    "dim(c)\n",
    "head(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cebc5de9-9af3-45a6-bd12-a3f97e9668bb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "spatial_data_subclones<-subset(spatial_data_sub, cells = rownames(c))\n",
    "\n",
    "seurat_colnames<-colnames(spatial_data_subclones)\n",
    "\n",
    "# Find the order of the BC values in the seurat_colnames\n",
    "bc_order <- match(rownames(c), seurat_colnames)\n",
    "\n",
    "# Reorder the columns of infercnv_sub_groupings based on the order\n",
    "c_reordered <- c[order(bc_order), ]\n",
    "spatial_data_subclones@meta.data$copykat_subclones <- c$x\n",
    "SpatialDimPlot(spatial_data_subclones, group.by = \"copykat_subclones\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d71b619-c205-4985-993c-d5bfc4d76af4",
   "metadata": {},
   "source": [
    "### Save the processed data for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4251b11a-d6db-4960-af78-718f61dcda37",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#saveRDS(spatial_data,\"spatial_data.rds\")\n",
    "#saveRDS(spatial_data_subclones,\"spatial_data_subclones.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85fea494-1ca2-46ad-a6ce-ca747874d44f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "<a id=\"SUMMARY\"></a>\n",
    "\n",
    "# SUMMARY\n",
    "\n",
    "Both of the discussed tools can be applied to scRNA-seq and spatial transcriptomics data\n",
    "\n",
    "- CopyKAT\n",
    "    - Cell-type annotation free tool to identify aneuploid (Tumor cells)\n",
    "    - To infer subclones of tumor\n",
    "    - Time consuming but easy to interpret\n",
    "    \n",
    "- InferCNV\n",
    "    - Mainly to predict CNV profiles of annotated tumor cells by comparing with the annotated normal cells and to identify subclones.\n",
    "    - Can also predict tumor cells when cell-type annotations are unknown\n",
    "    - Quick to run but not as easy as CopyKAT to interpret results\n",
    "    \n",
    "Other new tools to explore: 'SCEVAN' (https://www.nature.com/articles/s41467-023-36790-9)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "ir",
   "language": "R",
   "display_name": "R"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
