{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cell community identification using the hoodscanR package \n",
    "\n",
    "<img style=\"float: middle;\" src=\"../images/logo.png\">\n",
    "\n",
    "**This script has been generated by Feng Zhang and Onkar Mulay from the *Genomics and Machine Learning Lab*.** \n",
    "\n",
    "If you have any questions, please do not hesitate to contact us (feng.zhang@uq.edu.au or o.mulay@uq.edu.au)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [1 Introduction](#1-Introduction)<br>\n",
    "- [2 Cell community identification](#2-Cell-community-identification)<br>\n",
    "    - [2.1 Input data](#2.1-Input-data)<br>\n",
    "    - [2.2 Perform neighborhood analysis](#Perform-neighborhood-analysis)<br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1 Introduction \n",
    "### Why do we identify the cell community\n",
    "To understand the structure-function relationship of a tissue, the concept of tissue cellular neighborhoods (TCNs), niches or cell community has been proposed as a recurrent functional unit in which different cell types organize themselves to support tissue functions. \n",
    "\n",
    "Understanding the complexities of cell community is critical for advancing our knowledge of tissue biology and developing novel therapeutic strategies for various diseases.\n",
    "\n",
    "### What is CosMx data\n",
    "[CosMxâ„¢ Spatial Molecular Imager](https://nanostring.com/products/cosmx-spatial-molecular-imager/single-cell-imaging-overview/) is a fully-integrated multiomics single-cell spatial biology solution and enables researchers to investigate morphologically-intact tissues at unprecedented plex, resolution and cell numbers.\n",
    "\n",
    "### What tools can we use to run the cell community\n",
    "- [hoodscanR](https://github.com/DavisLaboratory/hoodscanR)\n",
    "- [NeighborhoodCoordination](https://github.com/nolanlab/NeighborhoodCoordination)\n",
    "- [CytoCommunity](https://github.com/huBioinfo/CytoCommunity)\n",
    "- [nichecompass](https://github.com/Lotfollahi-lab/nichecompass)\n",
    "- [SPIAT](https://github.com/TrigosTeam/SPIAT)\n",
    "- \n",
    "\n",
    "### What is hoodscanR\n",
    "hoodscanR is an user-friendly R package providing functions to assist cellular neighborhood analysis of any spatial transcriptomics data with single-cell resolution.\n",
    "\n",
    "All functions in the package are built based on the *SpatialExperiment* infrastructure, allowing integration into various spatial transcriptomics-related packages from Bioconductor. The package can result in cell-level neighborhood annotation output, along with funtions to perform neighborhood colocalization analysis and neighborhood-based cell clustering.\n",
    "\n",
    "![image.png](images/hoodscanR.jpg)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Cell community identification\n",
    "### 2.1 Input data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# library and set default parameter \n",
    "library(hoodscanR)\n",
    "library(SpatialExperiment)\n",
    "library(png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## This code will help speed up the plotting when running in base R on a Mac\n",
    "show_plot <- function(plot, filename = \"temp.png\", width = 1600, height = 1500, res = 260) {                                                  \n",
    "  temp_filename <- tempfile(fileext = \".png\")\n",
    "  grDevices::png(filename = temp_filename, width = width, height = height, res = res)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  img <- png::readPNG(temp_filename)\n",
    "  grid::grid.raster(img)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# set path and load data\n",
    "#data_dir <- '/scratch/project/stseq/Feng/package-vignette/temp-data'\n",
    "data_dir <- '/data/module2/data/'\n",
    "set.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# load and explore the CosMx data\n",
    "load(file.path(data_dir,'CosMx_Skin_Melanoma.RData'))\n",
    "print(dim(count))\n",
    "print(dim(spatial))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "count[1:5,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# plot the cell position on the tissue which cover 12 field of view (FOVs) \n",
    "spe <- SpatialExperiment(\n",
    "    assay = list(counts=count), \n",
    "    colData = spatial, \n",
    "    rowData = rownames(count),\n",
    "    spatialCoordsNames = c(\"sdimx\", \"sdimy\"))\n",
    "p1 <- plotTissue(spe, color = custom_clust, size = 0.5, alpha = 0.8)\n",
    "show_plot(p1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# take fov 12 for example\n",
    "count_fov <- count[,spatial$fov==12]\n",
    "spatial_fov <- spatial[spatial$fov==12,]\n",
    "dim(count_fov)\n",
    "dim(spatial_fov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(spatial_fov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "spe <- SpatialExperiment(\n",
    "    assay = list(counts=count_fov), \n",
    "    colData = spatial_fov, \n",
    "    rowData = rownames(count_fov),\n",
    "    spatialCoordsNames = c(\"sdimx\", \"sdimy\"))\n",
    "p2 <- plotTissue(spe, color = custom_clust, size = 1.5, alpha = 0.8)\n",
    "show_plot(p2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform neighborhood analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "spe <- readHoodData(spe, anno_col = \"custom_clust\")\n",
    "fnc <- findNearCells(spe, k = 100) # find 100 nearest cells for each cell\n",
    "lapply(fnc, function(x) x[1:10, 1:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# perform neighborhood analysis to genereate a matrix \n",
    "# with the probability of each cell associating with their 100 nearest cells.\n",
    "pm <- scanHoods(fnc$distance) #\n",
    "\n",
    "hoods <- mergeByGroup(pm, fnc$cells)\n",
    "p3 <- plotHoodMat(hoods, n = 10, hm_height = 5,hm_width = 8)\n",
    "show_plot(p3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "spe <- mergeHoodSpe(spe, hoods)\n",
    "spe <- calcMetrics(spe, pm_cols = colnames(hoods))\n",
    "spe <- clustByHood(spe, pm_cols = colnames(hoods), k = 10)\n",
    "p4 <- plotTissue(spe, color = clusters)\n",
    "show_plot(p4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# entropy and perplexity measure the mixture of neighborhood of each cell\n",
    "p5 <- plotTissue(spe, size = 1.5, color = perplexity) \n",
    "show_plot(p5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "p6 <- plotProbDist(spe, pm_cols = colnames(hoods), by_cluster = TRUE, plot_all = TRUE, show_clusters = as.character(seq(10)))\n",
    "show_plot(p6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# write.csv(spatial,file = 'spatial.csv',row.names=F,col.names=T,quote=F)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
